public/app/features/variables/adhoc/actions.ts

```diff
@@ -0,0 +1,164 @@
+  };
+};
+
+export const removeFilter = (uuid: string, index: number): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterRemoved(toVariablePayload(variable, index)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const addFilter = (uuid: string, filter: AdHocVariableFilter): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterAdded(toVariablePayload(variable, filter)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const setFiltersFromUrl = (uuid: string, filters: AdHocVariableFilter[]): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filtersRestored(toVariablePayload(variable, filters)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const changeVariableDatasource = (datasource: string): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const { editor } = getState().templating;
+    const variable = getVariable(editor.id, getState());
+
+    const loadingText = 'Adhoc filters are applied automatically to all queries that target this datasource';
+
+    dispatch(
+      changeVariableEditorExtended({
+        propName: 'infoText',
+        propValue: loadingText,
+      })
+    );
+    dispatch(changeVariableProp(toVariablePayload(variable, { propName: 'datasource', propValue: datasource })));
+
+    const ds = await getDatasourceSrv().get(datasource);
+    console.log('ds', ds);
+
+    if (!ds || !ds.getTagKeys) {
+      dispatch(
+        changeVariableEditorExtended({
+          propName: 'infoText',
+          propValue: 'This datasource does not support adhoc filters yet.',
+        })
+      );
+    }
+  };
+};
+
+export const initAdHocVariableEditor = (): ThunkResult<void> => dispatch => {
+  const dataSources = getDatasourceSrv().getMetricSources();
+  const selectable = dataSources.reduce(
+    (all: Array<{ text: string; value: string }>, ds) => {
+      if (ds.meta.mixed || ds.value === null) {
+        return all;
+      }
+
+      all.push({
+        text: ds.name,
+        value: ds.value,
+      });
+
+      return all;
+    },
+    [{ text: '', value: '' }]
+  );
+
+  dispatch(
+    changeVariableEditorExtended({
+      propName: 'dataSources',
+      propValue: selectable,
+    })
+  );
+};
+
+const createAdHocVariable = (options: AdHocTableOptions): ThunkResult<void> => {
+  return (dispatch, getState) => {
+    const model = {
+      ...cloneDeep(initialAdHocVariableModelState),
+      datasource: options.datasource,
+      name: filterTableName,
+      uuid: v4(),
+    };
+
+    const global = false;
+    const index = Object.values(getState().templating.variables).length;
```

Comment:
I think this index logic has spread in two or more places already, maybe we should promote it to an utility function?

Reply:
Sounds like a good plan, I will fix that.